S3Outputer.cc

```diff
@@ -0,0 +1,230 @@
+  using namespace std::string_literals;
+  auto p = currentProductStripes_.begin();
+  auto pi = index_.mutable_products()->begin();
+  for(; p != std::end(currentProductStripes_); ++p, ++pi) {
+    size_t offset = p->content().size();
+    size_t bufferNevents = p->offsets_size();
+
+    // first flush when we exceed min size and have an even divisor of eventFlushSize_
+    // subsequent flush when we reach productFlushSize
+    // always flush when we reach eventFlushSize_ (for buffers that never get big enough)
+    // flush if last call and we have something to write
+    if (
+        ((pi->flushsize() == 0) && (offset > productBufferFlushMinSize_) && (eventFlushSize_ % bufferNevents == 0))
+        || (bufferNevents == pi->flushsize())
+        || (bufferNevents == eventFlushSize_)
+        || (last && bufferNevents > 0)
+        )
+    {
+      if(verbose_ >= 2) {
+        std::cout << "product buffer for "s + std::string(pi->productname()) + " is full ("s + std::to_string(offset)
+          + " bytes, "s + std::to_string(bufferNevents) + " events), flushing\n" << std::flush;
+      }
+      objstripe::ProductStripe pOut;
+      pOut.mutable_offsets()->Reserve(bufferNevents);
+      pOut.mutable_content()->reserve(offset);
+      std::swap(*p, pOut);
+      std::string name = objPrefix_;
+      name += pi->productname();
+      name += std::to_string(eventGlobalOffset_ - bufferNevents);
+      iCallback.group()->run(
+        [this, name=std::move(name), pOut=std::move(pOut), callback=iCallback]() {
+          auto start = std::chrono::high_resolution_clock::now();
+          std::string finalbuf;
+          pOut.SerializeToString(&finalbuf);
+          conn_->put(name, std::move(finalbuf), [name=std::move(name), callback=std::move(callback)](S3Request* req) {
+              if ( req->status != S3Request::Status::ok ) {
+                std::cerr << "failed to write product buffer " << name << std::endl;
+              }
+            });
+          auto time = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now() - start);
+          parallelTime_ += time.count();
+        }
+      );
+      if ( pi->flushsize() == 0 ) {
+        pi->set_flushsize(bufferNevents);
+      }
+    }
+  }
+}
+
+void S3Outputer::flushEventStripe(TaskHolder iCallback, bool last) const {
+  if ( not last ) {
+    // all buffers should be empty because the sizes all evenly divide eventFlushSize_
+    for(auto& p : currentProductStripes_) {
+      assert(p->offsets_size() == 0);
+    }
+  }
+  index_.set_totalevents(eventGlobalOffset_);
+  if ( last and currentEventStripe_.events_size() == 0 ) {
+    return;
+  }
+  objstripe::EventStripe stripeOut;
+  stripeOut.mutable_events()->Reserve(eventFlushSize_);
+  std::swap(currentEventStripe_, stripeOut);
+  // TODO: are we sure writing to dest is threadsafe?
+  auto dest = index_.add_packedeventstripes();
```

Comment:
I don't think `dest` is anything more than a pointer in `index_` and only one thread will access `dest` ever, so this I think is safe

Reply:
`index_.add_packedeventstripes();` is extending an array of pointers and returning the newly added pointer. No two tasks will write to the data structure at the same pointer. On the other hand, later I serialize the whole thing which may read from that data in an invalid state. So I should ensure to only write out the events index at a safe time.